resources/guidelines/code/backward-compatibility.md

```diff
@@ -0,0 +1,751 @@
+#### TWIG Templates
+<table>
+	<tr>
+		<th width=260>Use Case</th>
+		<th width=120>Allowed?</th>
+		<th>Notes / Alternatives</th>
+	</tr>
+	<tr>
+		<td>Removing TWIG blocks.</td>
+		<td>🔴 NO</td>
+		<td>Use the deprecation workflow.<br><br>Code Example: [Deprecate TWIG block](#deprecate-twig-block)</td>
+	</tr>
+	<tr>
+		<td>Renaming TWIG blocks.</td>
+		<td>🔴 NO</td>
+		<td>Use the deprecation workflow. Create a new surrounding block with the new name and deprecate the old one. All variables which are defined in the scope of the old block, must be moved to the new surrounding block scope.<br><br>Code Example: [Rename TWIG block](#rename-twig-block)</td>
+	</tr>
+	<tr>
+		<td>Moving TWIG blocks within the same file.</td>
+		<td>⚪ PARTIAL</td>
+		<td>Only within the same scope / parent block.</td>
+	</tr>
+	<tr>
+		<td>Removing TWIG variables.</td>
+		<td>🔴 NO</td>
+		<td></td>
+	</tr>
+	<tr>
+		<td>Renaming TWIG variables.</td>
+		<td>🔴 NO</td>
+		<td>Create a new variable within the same scope and deprecate the old one.</td>
+	</tr>
+	<tr>
+		<td>Changing the value of TWIG variables</td>
+		<td>⚪ PARTIAL</td>
+		<td>The data type has to stay the same. Otherwise use the deprecation workflow.</td>
+	</tr>
+	<tr>
+		<td>Moving TWIG variable definitions to other TWIG blocks.</td>
+		<td>⚪ PARTIAL</td>
+		<td>Only when they are being moved higher up in the block scope.</td>
+	</tr>
+	<tr>
+		<td>Adding TWIG blocks which affect the scope of variable definitions.</td>
+		<td>🔴 NO</td>
+		<td></td>
+	</tr>
+	<tr>
+		<td>Moving template files to other directories.</td>
+		<td>🔴 NO</td>
+		<td></td>
+	</tr>
+</table>
+
+#### HTML
+<table>
+	<tr>
+		<th width=260>Use Case</th>
+		<th width=120>Allowed?</th>
+		<th>Notes / Alternatives</th>
+	</tr>
+	<tr>
+		<td>Removing HTML sections.</td>
+		<td>🔴 NO</td>
+		<td></td>
+	</tr>
+	<tr>
+		<td>Moving HTML sections within the same file.</td>
+		<td>⚪ PARTIAL</td>
+		<td>Only within the same TWIG Block.</td>
+	</tr>
+	<tr>
+		<td>Renaming of removing CSS selectors.</td>
+		<td>🔴 NO</td>
+		<td>Use the deprecation workflow.<br><br>Code Example: [Deprecate CSS selectors](#deprecate-css-selectors)</td>
+	</tr>
+</table>
+
+#### JavaScript
+<table>
+	<tr>
+		<th width=260>Use Case</th>
+		<th width=120>Allowed?</th>
+		<th>Notes / Alternatives</th>
+	</tr>
+	<tr>
+		<td>Renaming or removing JS services.</td>
+		<td>🔴 NO</td>
+		<td></td>
+	</tr>
+	<tr>
+		<td>Renaming or removing of JS plugins.</td>
+		<td>🔴 NO</td>
+		<td></td>
+	</tr>
+	<tr>
+		<td>Changing the public API of a JS plugin or service.</td>
+		<td>🔴 NO</td>
+		<td>Use the deprecation workflow.<br><br>Code Example: [Add new public function](#add-new-public-function)</td>
+	</tr>
+	<tr>
+		<td>Renaming methods of JS plugins or services.</td>
+		<td>🔴 NO</td>
+		<td>Use the deprecation workflow.<br><br>Code Example: [Rename a method](#rename-a-method)</td>
+	</tr>
+	<tr>
+		<td>Renaming or removing of JS events.</td>
+		<td>🔴 NO</td>
+		<td></td>
```

Comment:
Same here: Deprecation workflow wouldn't work here?

Reply:
See: https://github.com/shopware/docs/pull/278#discussion_r634978337

Reply:
I added a code example for this case and for other cases. For those ones which are missing we cannot really recommend an "easy/straight-forward" deprecation way because it too highly depends on the specific case (e.g. the route renaming, or renaming entire files etc.)

"No" in this document also doesn't mean that it's impossible/forbidden or something. It just means, that the developer cannot "just change" these things or simply put "deprecated" on it. It can be possible to make it backward compatible even when the list says "no". It's just not really possible to define a generic example which applies to all possible cases. Not sure how we can make that more clear. It should help the developer to find out what can be done easily and what needs closer attention. 

For example the "data props" in the Vue components. Those can contain dumb booleans or entire entity collections which get loaded via the API. Yes it is possible to deprecate those data props but they could be used in the template, in methods, in computed properties and so on and it would not be enough to say "Add deprecation to the property inside `data`" And the code example would almost be a small module.